More support for CENNZnet Extrinsics

HiruSylo opened the pull request:
Encoding an extrinsic panics if the FeeExchange version is not 1.
Change this to return an error instead.

HiruSylo committed:
Work in Progress (1)

DOES NOT WORK

HiruSylo committed:
Make basic extrinsics to work

HiruSylo committed:
Finish support for CENNZnet extrinsics

HiruSylo committed:
Update README

* Make README use the new name.
* Write about how this is a fork of the original upstream repo.

HiruSylo committed:
Update tests

HiruSylo committed:
Remove instance of code that can panic

Encoding an extrinsic panics if the FeeExchange version is not 1.
Change this to return an error instead.

HiruSylo committed:
Marshal u128 and u256 into JSON as hex strings

HiruSylo commented:
Also contains commits that add more support for CENNZnet including a fix for a bug where u128 and u256 numbers should be encoded as a hex string when used with the RPC.

HiruSylo committed:
Merge branch 'sylo/master' into add-cennznet-extrinsics

fishmonger45 commented on teste2e/author_submit_and_watch_extrinsic_test.go:
```diff
@@ -82,7 +82,7 @@ func TestAuthor_SubmitAndWatchExtrinsic(t *testing.T) {
 
 	var accountInfo types.AccountInfo
 	ok, err = api.RPC.State.GetStorageLatest(key, &accountInfo)
-	if err != nil || !ok {
+	if err != nil {
 		panic(err)
 	}
```
Does this `ok` go unused?

HiruSylo replied:
Yes, because it would error every time an account was used for the first time. I have no clue how their tests passed though. Maybe their runtime gives a default value? Another possibility is that they rely on the fact that go tests are sequential so they know for a fact that the test running in CI would be executed in order and thus Alice would have been guaranteed to have made a transaction before this.

fishmonger45 replied:
I mean we are not changing the way that  we are testing from go-substrate-rpc-client right? I'm fine to let this slide if you're confident that you don't need the `ok` check. I have come across the same issue with checking the balances of new accounts in that `ok` would be false. I have shorted the logic to defaulting this to being zero balance but didn't investigate very far. 

You can add a unit (`_`) instead of ok if you like.

HiruSylo replied:
> I mean we are not changing the way that we are testing from go-substrate-rpc-client right?

Yeah, no tests were changed

> You can add a unit (_) instead of ok if you like.

Sounds like a good fix for now.

I'd like to know why this differs for CENNZnet in the future though.

fishmonger45 replied:
I use slack as transient communication and github as an archive tool? Just adding from slack that this was an 'investigate later' sort of issue on the way substrate vs cennznet deem what is `ok`

HiruSylo committed:
Update test to be more correct

The ok variable was previously being binded but ignored.
This is because CENNZnet doesn't store any account info
if the account has never made a transaciton before.
To reflect this change, remove the assigning of that value to ok.
